cookbook/security/guard-api-key.rst

```diff
@@ -0,0 +1,489 @@
+.. index::
+    single: Security; Custom Request Authenticator
+
+How to Authenticate Users with an API Key using Guard
+=====================================================
+
+Nowadays, it's quite usual to authenticate the user via an API key (when developing
+a web service for instance). The API key is provided for every request and is
+passed as a query string parameter or via an HTTP header.
+
+Setting up the authenticator is just 3 steps:
+
+* :ref:`cookbook-guard-api-authenticator`
+* :ref:`cookbook-guard-api-configuration`
+* :ref:`cookbook-guard-api-user-provider`
+
+.. _cookbook-guard-api-authenticator:
+
+A) Create the Guard Authenticator
+---------------------------------
+
+Suppose you want to read an ``X-API-TOKEN`` header on each request and use
+that to authenticate the user. To do this, create a class that extends
+:class:`Symfony\\Component\\Security\\Guard\\AbstractGuardAuthenticator`
+(or which implements :class:`Symfony\\Component\\Security\\Guard\\GuardAuthenticatorInterface`)::
+
+    // src/AppBundle/Security/TokenAuthenticator.php
+    namespace AppBundle\Security;
+
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\Security\Core\Exception\AuthenticationCredentialsNotFoundException;
+    use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+    use Symfony\Component\Security\Core\Exception\AuthenticationException;
+    use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+    class TokenAuthenticator extends AbstractGuardAuthenticator
+    {
+        public function getCredentialsFromRequest(Request $request)
+        {
+            $token = $request->headers->get('X-AUTH-TOKEN');
+
+            // no token? Don't do anything :D
+            if (!$token) {
+                return;
+            }
+
+            return array(
+                'token' => $token,
+            );
+        }
+
+        public function authenticate($credentials, UserProviderInterface $userProvider)
+        {
+            $token = $credentials['token'];
+
+            // call a method on your UserProvider - see below for details
+            $user = $userProvider->loadUserByToken($token);
```

Comment:
Creating a specific `UserProvider` adds more complexity to the cookbook. There is easier solution for that. You can still use `return $userProvider->loadUserByUsername($token);` here
with the configuration:
```
providers:
        api_users:
            entity:
                class: AppBundle:User
                property: apiToken
```

On the other hand `loadUserByUsername($token)` may sound awkward, but this is another issue.